pkg/controller/transaction/controller.go

```diff
@@ -202,54 +208,78 @@ func (r *Reconciler) reconcileTransactionChangeValidating(ctx context.Context, t
 		return false, nil
 	}
 	return true, nil
-
 }
 
 func (r *Reconciler) reconcileTransactionChangeApplying(ctx context.Context, transaction *configapi.Transaction, change *configapi.TransactionChange) (bool, error) {
-	if transaction.Atomic {
-		// TODO: Apply atomic transactions here
-	} else {
-		for _, change := range change.Changes {
-			config, err := r.configurations.Get(ctx, change.TargetID)
-			if err != nil {
-				if !errors.IsNotFound(err) {
-					return false, err
-				}
-				return false, nil
+	transaction.Status.Sources = make(map[configapi.TargetID]*configapi.Source)
+	configs := make(map[configapi.TargetID]*configapi.Configuration)
+	for _, change := range change.Changes {
+		config, err := r.configurations.Get(ctx, change.TargetID)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				return false, err
 			}
-			updatedConfigValues := make(map[string]*configapi.PathValue)
-			transaction.Status.Sources = make(map[configapi.TargetID]*configapi.Source)
-			for _, pathValue := range config.Values {
-				pathValue.Index = transaction.Index
-				updatedConfigValues[pathValue.Path] = pathValue
-				transaction.Status.Sources[config.TargetID] = &configapi.Source{
-					Values: map[string]configapi.Index{
-						pathValue.Path: pathValue.Index,
-					},
-				}
+			return false, nil
+		}
+		configs[config.TargetID] = config
+
+		source := &configapi.Source{
+			Values: make(map[string]configapi.Index),
+		}
+		for _, changeValue := range change.Values {
+			pathValue, ok := config.Values[changeValue.Path]
+			if ok {
+				source.Values[pathValue.Path] = pathValue.Index
 			}
+		}
+		transaction.Status.Sources[config.TargetID] = source
+	}
 
-			config.Values = updatedConfigValues
-			config.Status.State = configapi.ConfigurationState_CONFIGURATION_PENDING
-			config.Status.TransactionIndex = transaction.Index
-			err = r.configurations.Update(ctx, config)
-			if err != nil {
-				if !errors.IsConflict(err) && !errors.IsNotFound(err) {
-					return false, err
-				}
-				return false, nil
+	// The source configurations must be stored prior to updating the target configurations
+	// otherwise they will be lost.
+	err := r.transactions.Update(ctx, transaction)
```

Comment:
We have to be really careful to make sure the information we’re storing about the source we’re changing is stored before we begin updating the target configurations. Otherwise, if we update a configuration and then crash, when we recover from the crash we’d be storing the updated target configuration as the source.

Reply:
Makes sense.